### DISCUSSION TITLE
Getting Started with Seeed Studio XIAO ESP32C3 | Seeed Studio Wiki

### DISCUSSION THREAD

**Discussion by giscus** [Seeed-Studio/wiki-documents#80] [Q&A]
# Getting Started with Seeed Studio XIAO ESP32C3 | Seeed Studio Wiki

Getting Started with Seeed Studio XIAO ESP32C3

https://wiki.seeedstudio.com/XIAO_ESP32C3_Getting_Started/

**Comment by gbakerLV** [comment]
I am trying to flash Circuitpython via the "Open Installer" on https://circuitpython.org/board/seeed_xiao_esp32c3/ not working.

Just want to manually accomplish.  Could you post the commands.  thx

**Reply by gbakerLV** [reply]
Okay, I downloaded the "flash download tool 3.9.4".
I went back and forth flashing your "ESP...1014.bin" and the CircuitPython. 
The tools said it was successful, but no CIRCUITPY drive.
Works okay with Ardiuino and your ".bin"
What am I missing

**Comment by vlkozl** [comment]
Url to `[PDF] ESP32C3 datasheet` is broken.

**Reply by MatthewJeffson** [reply]
Hi, I have messaged my colleague to fix it. Thank you for the notice!

**Comment by vlkozl** [comment]
Sorry, but where can I get info about how much power can I take from 5V, 3V3 and GPIO pins?
Thanks

**Reply by MatthewJeffson** [reply]
Hi, which kind of power are you looking for? Will this spec table can help you?

Parameter | Description
-- | --
Power | Circuit operating voltage: 3.3V@200mA
Charging current: 50mA/100mA
Input voltage (VIN): 5V
Deep Sleep Power Consumption | Deep Sleep Model: >44 μA
Wi-Fi Enabled Power Consumption | Active Model: <75 mA
| Modem-sleep Model: <25 mA
| Light-sleep Model: <4 mA
BLE Enabled Power Consumption | Modem-sleep Model: <27 mA
Light-sleep Model: <10 mA
Working Temperature | -40°C ~ 85°C

**Reply by vlkozl** [reply]
Hi, not quite that.
How many amps, for example, an attached sensor can drain from 5V, from 3v3?

**Reply by MatthewJeffson** [reply]
Hi, you got me with this. :D Would you mind help to test it out and tell me? I couldn't be more grateful :)

**Reply by vlkozl** [reply]
Hi, sorry I'm a newbie in DIY electronics, only getting use to it.

**Comment by osmoralv** [comment]
Hi, I'm try to using the arduino BLE example codes and when I click on "Verify", Arduino IDE launch this error (for "XIAO ESP32C3 as Bluetooth server" code):

Compilation error: 'class BLECharacteristic' has no member named 'getValue'; did you mean 'setValue'?

**Comment by xadonxander** [comment]
The file listed as factory does not seem to be the right file. I flashed it and now all I get is a pop up that says device not recognized every time I plug it into the computer.

**Comment by newtocodi** [comment]
Can I use two of this board in ESPnow without connecting extra device or chip

**Comment by snowbrown** [comment]
Does this ESP32 XIAO have DAC pins and I could not fine these DAC pins in this website.

**Reply by snowbrown** [reply]
This ESP32 XAIO has only one DAC port.

**Comment by PabloGonzalezOrtega** [comment]
How we can change the Tx power to increse the range of BLE? In ESP32 especifications it says that the max. power is arounr 20dBm. In Arduino IDE I can't see anything about that.

Thanks!

**Comment by Stebar59** [comment]
Hi, I'm trying to configure the XIAO ESP32C3 for usage with the 24GHz mmWave Human Static Presence Module Lite.
I've followed the instruction and created the ESPhome yaml file but when I try to download it I have following error:

INFO Reading configuration /config/esphome/movimento1.yaml...
INFO Generating C++ source...
INFO Compiling app...
Processing movimento1 (board: esp32-c3-devkitm-1; framework: espidf; platform: platformio/espressif32 @ 5.3.0)
--------------------------------------------------------------------------------
HARDWARE: ESP32C3 160MHz, 320KB RAM, 4MB Flash
 - framework-espidf @ 3.40404.0 (4.4.4) 
 - tool-cmake @ 3.16.9 
 - tool-ninja @ 1.10.2 
 - toolchain-esp32ulp @ 2.35.0-20220830 
 - toolchain-riscv32-esp @ 8.4.0+2021r2-patch5
Reading CMake configuration...
-- The C compiler identification is GNU 8.4.0
-- The CXX compiler identification is GNU 8.4.0
-- The ASM compiler identification is GNU
-- Found assembler: /data/cache/platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc
-- Check for working C compiler: /data/cache/platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc
-- Check for working C compiler: /data/cache/platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc -- broken
-- Configuring incomplete, errors occurred!
See also "/data/movimento1/.pioenvs/movimento1/CMakeFiles/CMakeOutput.log".
See also "/data/movimento1/.pioenvs/movimento1/CMakeFiles/CMakeError.log".

fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
CMake Error at /data/cache/platformio/packages/tool-cmake/share/cmake-3.16/Modules/CMakeTestCCompiler.cmake:60 (message):
  The C compiler

    "/data/cache/platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /data/movimento1/.pioenvs/movimento1/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/data/cache/platformio/packages/tool-ninja/ninja cmTC_919aa && [1/2] Building C object CMakeFiles/cmTC_919aa.dir/testCCompiler.c.obj
    [2/2] Linking C executable cmTC_919aa
    FAILED: cmTC_919aa 
    : && /data/cache/platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc -march=rv32imc  -nostartfiles -march=rv32imc --specs=nosys.specs CMakeFiles/cmTC_919aa.dir/testCCompiler.c.obj  -o cmTC_919aa   && :
    /data/cache/platformio/packages/toolchain-riscv32-esp/bin/../lib/gcc/riscv32-esp-elf/8.4.0/../../../../riscv32-esp-elf/bin/ld: /data/cache/platformio/packages/toolchain-riscv32-esp/bin/../libexec/gcc/riscv32-esp-elf/8.4.0/liblto_plugin.so: error loading plugin: /data/cache/platformio/packages/toolchain-riscv32-esp/bin/../libexec/gcc/riscv32-esp-elf/8.4.0/liblto_plugin.so: cannot open shared object file: No such file or directory
    collect2: error: ld returned 1 exit status
    ninja: build stopped: subcommand failed.
    
  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  /data/cache/platformio/packages/framework-espidf/tools/cmake/project.cmake:296 (__project)
  CMakeLists.txt:3 (project)

========================= [FAILED] Took 13.86 seconds =========================


Can someone please help me? I don't understand  the error messages.
Many thanks

Stefano

**Comment by MatthewJeffson** [comment]
Welcome to our product documentation platform! We value user feedback and contributions, so we couldn’t be more grateful that you can help us improving the documentation.

Hence, if you find obvious errors in the product documentation, or you have additional ideas and insights about using products, please [contribute](https://wiki.seeedstudio.com/Contribution-Guide) to us.

If you need **technical support**, you can email it to [fae@seeed.cc](mailto:fae@seeed.cc).

If you want to **discuss the product**, I recommend the [Seeed Studio Forum](https://forum.seeedstudio.com/) and [Seeed Studio Discord Channel](https://discord.gg/eWkprNDMU7)

The "Discussion channel" in Github here is a relatively new addition and may not receive much attention.

Nonetheless, we hope you find it helpful with the help of others. We will clean it up on a monthly basis.

**Comment by darekxan** [comment]
Does this board have a battery discharge cutoff voltage? Will it power down when battery gets discharged?

**Reply by numpad0** [reply]
Following this question. What exactly is this built in battery management situation? Is it from some unnamed chip or is this a clever software solution?

**Comment by BlackRose01** [comment]
Unfortunately, I recently had the problem that I always got the error "Write Timeout" after uploading my code. Flashing, as described in FAQ - Q3, also no longer worked. I got the same error. Therefore I had to do the following steps first:
1. connect pin 3.3V to D8
2. connect pin GND to D9

After that I could use the microcontroller again without any problems.

**Reply by BlackRose01** [reply]
Here is the source. Unfortunately only in German.
https://www.mikrocontroller.net/topic/562300

**Comment by msfujino** [comment]
I am having trouble uploading sketches.
I can upload sketches to two XIAOs without any problem, but I can't upload to one of them.
So I am trying to write the bootloader according to "Q3: I want to reflash the bootloader with factory firmware" on the Wiki, but it looks like the screenshot below and I can't write it.
It seems that the connection is made but the write timeout has occurred.

    DEBUG:['COM3', 'COM5']
    DEBUG:open com5 sucess
    DEBUG:select test bin
    DEBUG:Flash
    DEBUG:['C:/ESP_TOOLS/EspRFTestTool_v3.6_Manual/ESP32-C3_RFTest_108_2b9b157_20211014.bin']
    DEBUG:Com is closed ser0
    DEBUG:close com sucessfully
    ERROR:Write timeout
    WARNING:Sync fail...

I have already confirmed the following
    Connect to USB while holding down the boot switch
    Remove the com port and reconnect it
    Update the driver for the com port

Is there anything else I can try?

**Reply by BlackRose01** [reply]
Try this one:
[https://github.com/Seeed-Studio/wiki-documents/discussions/80#discussioncomment-9009831](https://github.com/Seeed-Studio/wiki-documents/discussions/80#discussioncomment-9009831)

**Comment by agahvcn** [comment]
"Is there a pin I can use to trigger a MOSFET in deep sleep mode?"

**Reply by nfs0619** [reply]
You can learn how to configure GPIO pins during deep sleep by reading the following documentation:

- [ESP-IDF Programming Guide: Sleep Modes (ESP32-C3)](https://docs.espressif.com/projects/esp-idf/en/v5.4/esp32c3/api-reference/system/sleep_modes.html)

- [ESP-IDF Programming Guide: Sleep Modes (ESP32)](https://docs.espressif.com/projects/esp-idf/en/v5.4/esp32/api-reference/system/sleep_modes.html)

**Comment by ruehlchris** [comment]
Hi,
Running in an issue with the external 5V supply. The Led show charging like plug in the USB
and the 3v3 Pin is up too. But the board is not starting.  I plug the USB - with and without CDC enabled 
and everything comes alive.
Is the 5V input only to charge a battery and supply 3v3?
Please clarify, and if so - you may update the description in the Getting Started.. 

My project I use the esp32c3 to replace a toy's RC car malfunction remote control with GATT-server/client.

Chris

**Reply by msfujino** [reply]
Is there a "while(!Serial);" in your sketch? 
If so, remove it and it will work without USB.

**Reply by ruehlchris** [reply]
no such while() in my code. That was my first thing thinking that the code hang in a Terminal ready wait loop.

**Reply by msfujino** [reply]
According to the schematic, the 5V pin and the USB 5V are connected internally, so unless the two are connected at the same time, the problem should not occur.
Also, the ESP32C3 consumes a lot of current when the radio is enabled (over 200mA peak), and I am concerned about whether the external 5V power supply can supply the current properly.

**Reply by ruehlchris** [reply]
I got the supply from 3V power line from the RC car's PCBA and have a 2A power up stepper to 5.1V. The supply is stable.
I'm checking gpio9 and gpio2 for SPI boot must have pull up , have a weak pull up soldered from 3v3 -> gpio2 (D0)
gpio9 is untouched and 3v3.  Still no boot.

**Reply by ruehlchris** [reply]
Found it! It's a problem with the power-rail on the PCBA from the RC-car.  There is current limiter and when I start the board the output voltage drops to 1.8V on that rail.
I need to fix this. I get back to you.

**Comment by ruehlchris** [comment]
OK,
short update. I have build the Remote control unit with a Li-Po battery and it works flawless with & w/o CDC enabled.
I waiting now for a MPS (input 7.5~28V output 5V/1.6A) which then connect to the unlimited 9V rail of the RC-Car's supply.

**Reply by nfs0619** [reply]
Thank you for sharing! We’re excited to welcome you to contribute to our project.

If you're interested in learning more about how to get involved or contribute, feel free to check out the links below for more information:

- [Contributor Introduction](https://wiki.seeedstudio.com/contributors/)
- [Contributor Project Page](https://github.com/orgs/Seeed-Studio/projects/6/views/1?pane=issue&itemId=30957479)

We look forward to more interactions and collaboration with you!

**Comment by jkyumekamui** [comment]
Does ESP32 C3 really have FPU?
I didn't see anything about FPU in [the ESP32 C3 datasheet](https://www.espressif.com/sites/default/files/documentation/esp32-c3_datasheet_en.pdf) and it is mentioned in 4.1.1 in [the ESP32 one](https://www.espressif.com/sites/default/files/documentation/esp32_datasheet_en.pdf). According to [API guides of ESP32 C3](https://docs.espressif.com/projects/esp-idf/en/stable/esp32c3/api-guides/performance/speed.html), floating point arithmetic are emulated in software and are very slow.

**Reply by nfs0619** [reply]
Thank you very much for your feedback, and we apologize for the confusion. The XIAO ESP32 C3 does not have a built-in FPU. We have already corrected the issue in the wiki documentation, and we truly appreciate you pointing this out. As you mentioned, floating point operations are emulated in software on the XIAO ESP32 C3, which can lead to slower performance. We appreciate your understanding and support!

**Comment by GeGou** [comment]
Hi, I am trying to co next a neo 6m gos module and a sim800l module. Gos module is connected to TX RX pins and works fine, but I cannot make gprs module to work using pins D9 D10. I have already try the same in a lolin esp32 c3 board and there works, I use the same code with xiao and doesn't work. Any suggestion please;

**Reply by nfs0619** [reply]
Hi, if you want to use D9 and D10 as a serial port to communicate with the SIM800L module, you need to use the `SoftwareSerial` library. Please make sure your code correctly utilizes this library.  

If you have any further questions, feel free to contact our technical support at **[techsupport@seeed.io](mailto:techsupport@seeed.io)**.

**Reply by GeGou** [reply]
Hi , thank you for the reply , i have try what i saw from this thread https://forum.seeedstudio.com/t/xiao-esp32c3-multiple-software-serial-ports/291155/38 but still cannot make it run. i have paste the simple code i am trying to run. I testes it with lolin c3 super mini board and it works, means that passwords, usernames and apns are correct. If you can make some changes to test it again i will be very thankful. Comments are in greek, that maybe be a problem.
[New Text Document.txt](https://github.com/user-attachments/files/19133486/New.Text.Document.txt)

**Reply by GeGou** [reply]
Finally I solve the problem by using hardwareSerial on pins gpio6 and gpio7 if I remember correct.

**Comment by Shivang1303** [comment]
i am getting an issue when i am trying to power up my xiao esp32c3 and esp32c6 using a 5V power supply
i have tested current limit from 40mA to 120mA, every time i try to power it via VUSB pin, the current drawn from power supply fluctuates, and i am not able to detect it in ble mode

i have even tried using a diode between the vusb pin and power supply, and when i was powering it with USB cable, i checked the current drawn was around 69mA for c6 and 76mA for c3

please help me troubbleshoot this issue

**Reply by Shivang1303** [reply]
and also, the red led constantly blinks when powering with external power supply, but both the boards work perfectly with usb cable

**Reply by nfs0619** [reply]
Hello, we do not recommend powering the board directly via the VUSB pin, as this may cause unstable current issues. Additionally, if a Type-C cable is connected at the same time, it could damage the board. We recommend using the **Battery Connector** pin on the back of the board for a safer and more stable power supply.

**Reply by Shivang1303** [reply]
i used the battery connector pins on the back of the board, C3 takes initially 160mA to turn on, in ble mode connected to a sensor and sending data, and then after 1s it takes a stable 75-76mA current. c6 takes initially about 250mA and then 69mA afterwards. is there any way i can reduce or configure this initial current required?

**Reply by nfs0619** [reply]
Hello, we’re sorry, but we do not provide a way to reduce or configure the initial current. The Xiao ESP32C3 and ESP32C6 require a higher inrush current during startup, which is a normal phenomenon. We recommend using a power supply capable of providing sufficient inrush current to ensure stable device startup.

**Comment by Shivang1303** [comment]
Ok, thanks for the reply.

**Comment by pop-aruaru** [comment]


**Reply by nfs0619** [reply]
Hello,

The term "built-in" in the documentation means that the functionality is integrated within the main chip itself, rather than using an external, separately soldered IC.

**Reply by nfs0619** [reply]
Yes, U1 in the C3 schematic is a battery charging IC, specifically the ETA4054S2F. I understand your point about how the "built-in" description might be misleading when compared to the XIAO-nRF52840, which explicitly states BQ25101. We’ll work on improving the wording in our wiki to make this clearer. Thanks for your valuable feedback!

**Comment by kotbehemot53** [comment]
Hi, I have 2 questions.

1. What if I power the board via the 5V pin (using an external USB PSU) and don't use a diode, as suggested above?
2. Is the 5V pin on the board directly connected to the 5V USB rail? If I power the board via the USB-C, can I power external devices via the 5V pin without worrying that the current draw would affect the ESP?

**Reply by kotbehemot53** [reply]
Question 2. relates to the built-in USB-C connector and is of course assuming that the PSU used can handle the combined current draw of the board and the external devices connected to the 5V pin.

**Reply by nfs0619** [reply]
Thank you for your inquiry.

If you don't use a diode, it may cause current backflow. We don't recommend powering the board through the 5V pin, as it is typically used for output rather than input.

Additionally, the 5V pin is directly connected to the USB 5V rail. In theory, if you're powering the board via USB-C, you can power external devices through the 5V pin. However, we recommend ensuring that the power consumption of external devices isn't too high to avoid excessive current.

If you need more detailed information, I recommend reaching out to our technical support at [techsupport@seeed.io](mailto:techsupport@seeed.io).

**Reply by kotbehemot53** [reply]
"If you don't use a diode, it may cause current backflow."
The backflow scenario would only apply to a situation, where there is power connected to the USB-C connector and to the 5V pin at the same time, right? If they are never used at the same time, the diode is redundant, am I correct?

"the 5V pin is directly connected to the USB 5V rail" "we recommend ensuring that the power consumption of external devices isn't too high to avoid excessive current"
Why is overcurrent a concern (assuming that the PSU can handle it)? If the pin is directly connected to the 5V rail, without any components in-between, the only issue might be with the trace connecting the 5V pin to the USB-C connector being too narrow?

I'm asking all this, because I designed a XIAO ESP32C3-powered board that drives a bunch of LED lights. My board has an external 5V power connector, that powers the XIAO (via the 5V pin) and the LEDs. The XIAOs GPIO pins are hooked to BJT NPN transistors that cut or connect the LEDs' cathodes to GND. I'm wondering if the external power connector is at all necessary, or if the whole thing (including the LEDs) could be powered directly from the XIAO built-in USB-C. If the 5V pin is connected to the 5V USB rail directly, it seems that using the external power connector is electrically equivalent to using the built-in USB-C (except for a possibility that the trace connecting USB-C to the 5V pin is too narrow and might overheat when all the current goes through it).

**Reply by nfs0619** [reply]
Thank you for the follow-up! You're correct about the backflow issue – it only occurs if both USB-C and the 5V pin are powered simultaneously. If only one is active, the diode isn't necessary.

The concern about overcurrent is indeed related to the PCB trace width. If the trace is too narrow, it could cause overheating or damage.

In theory, you can power both the XIAO and LEDs from the USB-C port. The key is ensuring that the total current draw doesn’t exceed the USB-C power supply's capacity, and that the PCB trace can handle it safely.

If you need more information, I recommend reaching out to our technical support at [techsupport@seeed.io](mailto:techsupport@seeed.io).

**Reply by kotbehemot53** [reply]
Thank you for the replies! I now know all I need to know :-)

**Comment by abhishekgyani** [comment]
how to run jq6500 mp3 module with esp32 c3 using a push button ?  can anyone share a program ?

**Reply by nfs0619** [reply]
You can refer to this document to help you write your code:  
[https://wiki.seeedstudio.com/XIAO_ESP32C3_Pin_Multiplexing/](https://wiki.seeedstudio.com/XIAO_ESP32C3_Pin_Multiplexing/)

You may also check our forum to see if someone has already solved a similar issue:  
[https://forum.seeedstudio.com/](https://forum.seeedstudio.com/)

**Comment by Mah819** [comment]
Hi, would it be possible to power the board with a 6~6.6v Li-Po battery through the battery pads?

**Reply by nfs0619** [reply]
Hi! The XIAO ESP32C3 is designed to be powered by a 3.7V lithium battery. Supplying 6~6.6V directly through the battery pads is not recommended, as it may damage the board.

**Reply by Mah819** [reply]
Thank you for the reply

**Reply by Mah819** [reply]
Hi, would a 5v battery instead be possible? Or do I have to power the board strictly with a 3.7v power supply through the battery pads and if I were to power it with a 5v power supply, I would have to do it through the 5v pins?

**Reply by nfs0619** [reply]
Hi! A 5V battery may also damage the board if connected through the battery pads. For power supply guidance, please refer to this section: [Battery Usage – XIAO ESP32C3 Wiki](https://wiki.seeedstudio.com/XIAO_ESP32C3_Getting_Started/#battery-usage)

**Reply by Mah819** [reply]
OK, thank you for the reply

**Comment by bigjoejc** [comment]
TL;DR:
My Seeed Studio XIAO ESP32-S3 Sense appears to be defective. Despite extensive troubleshooting, it will not reliably enter bootloader mode or accept firmware uploads. Multiple known-good cables, ports, PCs, Arduino IDE versions, and clean example sketches were tested. The board powers on, but uploads consistently fail with “Failed to connect to ESP32: No serial data received”. This strongly suggests a hardware fault (USB-CDC / boot circuitry).

Details (brief):

Board: XIAO ESP32-S3 Sense (OV2640, 8MB PSRAM / 8MB Flash)

Arduino IDE + esp32 core up to date

Correct board selected (XIAO ESP32-S3 / ESP32-S3 settings tested)

Multiple PCs, USB cables, and ports tested

Boot button / reset sequences tried per documentation

Clean sketches (WiFi / minimal examples)

Board enumerates, but will not enter flashing mode reliably

Upload fails consistently via IDE and esptool

Requesting replacement or RMA, as this appears to be a hardware defect.

**Reply by nfs0619** [reply]
Hello, thank you for your detailed feedback. For this issue, we recommend contacting our technical support team directly at **[techsupport@seeed.io](mailto:techsupport@seeed.io)**. Please include relevant details such as the troubleshooting steps you have already performed, your setup, and any error messages you encountered. Our technical support team will be able to provide faster and more detailed assistance.

**Comment by Johannes-Obermaier** [comment]
Link to the  KiCad Footprint library is broken

**Reply by nfs0619** [reply]
Thank you for the reminder. We have fixed the link.
